technical/binary_tree/c++/binary_tree.h

```diff
@@ -0,0 +1,295 @@
+#include <stack>
+#include <iostream>
+
+template <typename T>
+// Expects type T to have an operator overload for <. 
+class BinaryTree {
+  friend class Iterator;
+private:
+  struct node {
+    T data;
+    node* left;
+    node* right;
+    node* parent;
+    node(T val, node* parent): data(val), left(nullptr), right(nullptr), parent(parent) {}
+  };
+  node* root;
+  void delete_helper(node*);
+  void add_helper(T, node* &);
+  bool find_helper(T, node*);
+  void print_helper(node*);
+  node*& find_node(T, node*&);
+  node* in_order_successor(node*);
+  node* in_order_predecessor(node*);
+  void remove_helper(node*&);
+public:
+  // does an in-order traversal of the tree
+  // modifying the tree invalidates the Iterator
+  class Iterator {
+    friend class BinaryTree;
+  private:
+    std::stack<node*> nodes;
+  public:
+    Iterator();
+    Iterator(const Iterator&); //cpy-ctor
+    Iterator operator=(const Iterator&);
+    Iterator operator++();
+    T& operator*(); // get data at this node
+    bool operator!=(const Iterator&);
+    bool operator==(const Iterator&);
+  };
+  BinaryTree();
```

Comment:
No copy constructor? Also, if you want to be really thorough, you can also implement the move constructor and move assignment